Source/WebCore/page/NavigatorLoginStatus.cpp

```diff
@@ -47,32 +58,43 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
     return "NavigatorLoginStatus"_s;
 }
 
-void NavigatorLoginStatus::setLoggedIn(Navigator& navigator, Ref<DeferredPromise>&& promise)
+void NavigatorLoginStatus::setLoginStatus(Navigator& navigator, const String& status, Ref<DeferredPromise>&& promise)
 {
-    NavigatorLoginStatus::from(navigator)->setLoggedIn(WTFMove(promise));
-}
-
-void NavigatorLoginStatus::setLoggedOut(Navigator& navigator, Ref<DeferredPromise>&& promise)
-{
-    NavigatorLoginStatus::from(navigator)->setLoggedOut(WTFMove(promise));
+    NavigatorLoginStatus::from(navigator)->setLoginStatus(status, WTFMove(promise));
 }
 
 void NavigatorLoginStatus::isLoggedIn(Navigator& navigator, Ref<DeferredPromise>&& promise)
 {
     NavigatorLoginStatus::from(navigator)->isLoggedIn(WTFMove(promise));
 }
 
-void NavigatorLoginStatus::setLoggedIn(Ref<DeferredPromise>&& promise)
+void NavigatorLoginStatus::setLoginStatus(const String& status, Ref<DeferredPromise>&& promise)
 {
-    if (m_navigator.cookieEnabled())
-        promise->resolve();
+    if (status == WTF::String::fromLatin1("logged-in"))
+        m_loginStatus = IsLoggedIn::Yes;
+    else if (status == WTF::String::fromLatin1("logged-out"))
+        m_loginStatus = IsLoggedIn::No;
     else
         promise->reject();
-}
 
-void NavigatorLoginStatus::setLoggedOut(Ref<DeferredPromise>&& promise)
-{
-    promise->resolve();
+    Document* document = m_navigator.document();
```

Comment:
We should use a smart pointer here for Document. We also probably need a null check.
```suggestion
    RefPtr document = m_navigator.document();
```